make-rules/environment.mk

```diff
@@ -48,6 +52,63 @@ component-environment-prep::
 		{ echo "Adding required packages to build environment..."; \
 		  $(PFEXEC) /usr/bin/pkg install --accept -v $(REQUIRED_PACKAGES:%=/%) || [ $$? -eq 4 ] ; }
 
+ZONENAME_PREFIX = bz
+ZONENAME_ID = $(shell echo "$(WS_TOP)" | sha1sum | cut -c0-7)-$(COMPONENT_NAME)
+ZONENAME = $(ZONENAME_PREFIX)-$(ZONENAME_ID)
+
+component-zone-template:
+	$(call separator-line)
+	$(call separator-line,Create template zone)
+	USER_ID=$$(id -u) && \
+	$(PFEXEC) $(ZONE) --prefix $(ZONENAME_PREFIX) create-template -u $${USER} -i $${USER_ID}
+
+component-zone-build:
+	$(call separator-line)
+	$(call separator-line,Create $(ZONENAME))
+	$(PFEXEC) $(ZONE) --prefix $(ZONENAME_PREFIX) spawn-zone --id $(ZONENAME_ID)
+	$(call separator-line,Boot $(ZONENAME))
+	@while $$(true); do \
+		echo "Waiting for zone $(ZONENAME) to boot..."; \
+		$(PFEXEC) /usr/sbin/zlogin -l $${USER} $(ZONENAME) \
+				/bin/true >/dev/null 2>&1 && break; \
+		sleep 10; \
+	done
+	# FIXME:
+	# - remove once we figure out a better way how to enable zoneproxy-client inside the nlipkg brand
+	@while $$(true); do \
+  		echo "Waiting for $(ZONENAME) config repository.."; \
+  		$(PFEXEC) /usr/bin/svcs -z $(ZONENAME) -a >/dev/null 2>&1 && break; \
+  		sleep 10; \
+  	done
+	# We need to create door inside  after zone-proxy-client is running
+	$(call separator-line,Configure IPS for $(ZONENAME))
+	$(PFEXEC) /usr/lib/zones/zoneproxy-adm $(ZONENAME)
+	@while $$(true); do \
+  		echo "Waiting for zoneproxyd to be ready.."; \
+  		PROXY_PID=$$(/usr/bin/svcs -p svc:/application/pkg/zones-proxyd:default | \
+  			nawk '$$0 ~ /zoneproxyd/ {print $$2}') && \
+  			$(PFEXEC) /usr/bin/pfiles $${PROXY_PID} | \
+  			$(GNU_GREP) $(ZONENAME) >/dev/null 2>&1 && break; \
+  		sleep 10; \
+  	done
+	$(PFEXEC) /usr/sbin/svcadm -z $(ZONENAME) \
+		enable svc:/application/pkg/zones-proxy-client:default
+	ZONEROOT="$$(/usr/sbin/zoneadm -z $(ZONENAME) list -p | cut -d: -f4)/root" && \
+		$(PFEXEC) /usr/bin/pkg -R $${ZONEROOT} set-property use-system-repo True && \
+	while $$(true); do \
+		echo "Waiting for sysrepo to be ready..." && \
+		$(PFEXEC) /usr/bin/pkg -R $${ZONEROOT} publisher | \
+		$(GNU_GREP) syspub >/dev/null 2>&1 && break; \
+		sleep 10; \
+	done
+	$(call separator-line,Build in $(ZONENAME))
+	$(PFEXEC) /usr/sbin/zlogin -l $${USER} $(ZONENAME) \
+		"cd $(COMPONENT_DIR); gmake install"
```

Comment:
why always do install, not publish (or custom target)? What is expected use case?

Reply:
The expected use case is to run `zone-build`, which will do `build install` in the zone. We can discuss changing `build install` to `publish`. However, the issue is that zone has no networking, so `publish` will fail. It's expected to run `download` in GZ with working networking. Once built, we can run gmake publishing globally.

I mean I am open to changing and discusisng this, but would not like to complicate setup with working networkingin the zones.

Reply:
Hi xen0l, what to do next?